libflux/src/semantic/analyze.rs

```diff
@@ -0,0 +1,2604 @@
+        Ok(FunctionExpr {
+            loc: &expr.base.location,
+            type_info: self.type_info(),
+            params,
+            body,
+        })
+    }
+
+    fn analyze_function_params<'a>(
+        &mut self,
+        props: &'a Vec<ast::Property>,
+    ) -> Result<Vec<FunctionParameter<'a>>> {
+        // The iteration here is complex, cannot use iter().map()..., better to write it explicitly.
+        let mut params: Vec<FunctionParameter> = Vec::new();
+        let mut piped = false;
+        for prop in props {
+            let id = match &prop.key {
+                ast::PropertyKey::Identifier(id) => Ok(id),
+                _ => Err("function params must be identifiers".to_string()),
+            }?;
+            let key = self.analyze_identifier(&id)?;
+            let mut default: Option<Expression> = None;
+            let mut is_pipe = false;
+            match &prop.value {
+                Some(expr) => match expr {
+                    ast::Expression::PipeLit(_) => {
+                        if piped {
+                            return Err("only a single argument may be piped".to_string());
+                        } else {
+                            piped = true;
+                            is_pipe = true;
+                        };
+                    }
+                    e => default = Some(self.analyze_expression(&e)?),
+                },
+                None => (),
+            };
+            params.push(FunctionParameter {
+                loc: &prop.base.location,
+                is_pipe,
+                key,
+                default,
+            });
+        }
+        Ok(params)
+    }
+
+    fn analyze_function_body<'a>(
+        &mut self,
+        body: &'a ast::FunctionBody,
+    ) -> Result<FunctionBody<'a>> {
+        match &body {
+            ast::FunctionBody::Expr(expr) => {
+                Ok(FunctionBody::Expr(self.analyze_expression(&expr)?))
+            }
+            ast::FunctionBody::Block(block) => Ok(FunctionBody::Block(self.analyze_block(&block)?)),
+        }
+    }
+
+    fn analyze_block<'a>(&mut self, block: &'a ast::Block) -> Result<Block<'a>> {
+        let stmts = block
+            .body
+            .iter()
+            .map(|s| self.analyze_statement(s))
+            .collect::<Result<Vec<Statement>>>()?;
+        match &stmts.get(stmts.len() - 1) {
+            Some(Statement::Return(_)) => Ok(Block {
+                loc: &block.base.location,
+                body: stmts,
+            }),
+            _ => Err("missing return statement in block".to_string()),
+        }
+    }
+
+    fn analyze_call_expression<'a>(&mut self, expr: &'a ast::CallExpr) -> Result<CallExpr<'a>> {
+        let callee = self.analyze_expression(&expr.callee)?;
+        // TODO(affo): I'd prefer these checks to be in ast.Check().
+        if expr.arguments.len() > 1 {
+            return Err("arguments are more than one object expression".to_string());
+        }
```

Comment:
This can be removed now.

Reply:
Actually never mind. I misread this. Although I would definitely prefer that this be changed in the AST. But not necessary for this PR.